First pass - Documentation improvement (#548)
kubernetes-sigs/kueue: docs/concepts/resource_flavor.md

> # ResourceFlavor object

Reviewer: ```suggestion
# Resource Flavor
```

for consistency with the other pages. Or maybe they should be titled `ClusterQueue` and `LocalQueue` instead? like  https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/

Resources in a cluster are typically not homogeneous. Resources could differ in:

> - pricing and availability (ex: spot vs on-demand VMs)

Reviewer 2: ```suggestion
- pricing and availability (for example, spot versus on-demand VMs)
```

> - architecture (ex: x86 vs ARM CPUs)

Reviewer 2: ```suggestion
- architecture (for example, x86 versus ARM CPUs)
```

> - brands and models (ex: Radeon 7000 vs Nvidia A100 vs T4 GPUs)

Reviewer 2: ```suggestion
- brands and models (for example, Radeon 7000 versus Nvidia A100 vs T4 GPUs)
```

A ResourceFlavor is an object that represents these resource variations and

> allows you to associate them with node labels and taints.

Reviewer: ```suggestion
allows you to associate them with cluster Nodes through labels and taints.
```

> **Note**: If your cluster is homogeneous, you can use an [empty ResourceFlavor](#empty-resourceflavor)

Reviewer 2: ```suggestion following the initial paragraph terminology
**Note**: If the resources in your cluster are homogeneous, you can use an [empty ResourceFlavor](#empty-resourceflavor)
```

instead of adding labels to custom ResourceFlavors.

A sample ResourceFlavor looks like the following:

```yaml
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: spot
nodeSelector:
  instance-type: spot
taints:
- effect: NoSchedule
  key: spot
  value: "true"
```

> You can use the `.metadata.name` to reference a ResourceFlavor from a

Reviewer 2: ```suggestion
You can use the `.metadata.name` field to reference a ResourceFlavor from a
```

> ClusterQueue in the `.spec.resources[*].flavors[*].name` field.

Reviewer: link to ClusterQueue

## ResourceFlavor labels

**Requires Kubernetes 1.23 or newer**

To associate a ResourceFlavor with a subset of nodes of your cluster, you can
configure the `.nodeSelector` field with matching node labels that uniquely identify
the nodes. If you are using [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)

> (or equivalent controllers), make sure it is configured to add those labels when

Reviewer 2: ```suggestion
(or equivalent controllers), make sure the controller is configured to add those labels when
```

I'm just assuming "it" refers to "the controller". I highly recommend you change "it" for the subject name to avoid confusion.

adding new nodes.

> To guarantee that the workload Pods run on the nodes associated to the flavor
> that Kueue decided that the workload should use, Kueue performs the following
> steps:

Reviewer 2: ```suggestion
To guarantee that the workload Pods run on the nodes associated to the flavor
that Kueue selected, Kueue performs the following
steps:
```

Reviewer 2: This is just to make the sentence more concise

1. When admitting a workload, Kueue evaluates the
   [`.nodeSelector`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector)
   and [`.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity)
   fields in the PodSpecs of your [Workload](workload.md) against the

>    ResourceFlavor labels.

Reviewer: Add:

ResourceFlavors that don't match the node affinity of the Workload cannot be assigned to a Workload's podSet.

We should probably add examples about this, but it could be a follow up.

2. Once the workload is admitted, Kueue adds the ResourceFlavor labels to the

>   `.nodeSelector` of the underlying workload Pod templates, if the workload
>    didn't specify them already.

Reviewer 2: ```suggestion
  `.nodeSelector` of the underlying workload Pod templates. This occurs if the workload
   didn't specify the ResourceFlavor labels already.
```

   For example, for a [batch/v1.Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/),
   Kueue adds the labels to the `.spec.template.spec.nodeSelector` field. This

>    guarantees that the Workload's Pods can only be scheduled on the nodes

Reviewer 2: ```suggestion
   guarantees that the workload's Pods can only be scheduled on the nodes
```

>    targeted by the flavor that Kueue assigned to the Workload.

Reviewer 2: ```suggestion
   targeted by the flavor that Kueue assigned to the workload.
```

Reviewer: When referring to API objects, the kubernetes documentation uses CamelCase.

Author: Done

Author: @cortespao could you do a quick pass? I just reviewed and tried to differentiate between Workload and workload... It's in the last commit.

Reviewer 2: I did a quick pass and all seems right. Thanks, Roberto 😄

## ResourceFlavor taints

To restrict the usage of a ResourceFlavor, you can configure the `.taints` field

> with taints.

Reviewer: These taints should typically match the taints of the Nodes associated with this ResourceFlavor.

Taints on the ResourceFlavor work similarly to [node taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).

> For Kueue to admit a workload to use the ResourceFlavor, the PodSpecs in the

Reviewer 2: Maybe "allow" is better than "admit" in this context?

Reviewer: There are two related words:
Admission and assignment. During admission, a workload is assigned flavors.
Not sure how to rephrase the sentence, though.

> workload should have a toleration for it. As opposed to the behavior for
> [ResourceFlavor labels](#resourceflavor-labels), Kueue does not add tolerations
> for the flavor taints.

Reviewer 2: ```suggestion
workload should have a toleration for it. In contrast, Kueue does not add tolerations
for the flavor taints with [ResourceFlavor labels](#resourceflavor-labels).
```

Please, disregard if my suggestion affects the accuracy.

Reviewer: Maybe we can remove the mention of labels here.

## Empty ResourceFlavor

If your cluster has homogeneous resources, or if you don't need to manage
quotas for the different flavors of a resource separately, you can create a
ResourceFlavor without any labels or taints. Such ResourceFlavor is called an

> empty ResourceFlavor and its sample looks like the following:

Reviewer: ```suggestion
empty ResourceFlavor and its definition looks like the following:
```

allows you to associate them with cluster nodes through labels and taints.

**Note**: If the resources in your cluster are homogeneous, you can use 
an [empty ResourceFlavor](#empty-resourceflavor) instead of adding labels to custom ResourceFlavors.

A sample ResourceFlavor looks like the following:

```yaml
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: spot
nodeSelector:
  instance-type: spot
taints:
- effect: NoSchedule
  key: spot
  value: "true"
```

You can use the `.metadata.name` field to reference a ResourceFlavor from a
[ClusterQueue](/docs/concepts/cluster_queue.md) in the `.spec.resources[*].flavors[*].name` field.

## ResourceFlavor labels

**Requires Kubernetes 1.23 or newer**

To associate a ResourceFlavor with a subset of nodes of your cluster, you can
configure the `.nodeSelector` field with matching node labels that uniquely identify
the nodes. If you are using [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
(or equivalent controllers), make sure that the controller is configured to add those labels when
adding new nodes.

> To guarantee that the Workload's Pods run on the nodes associated to the flavor

Reviewer 2: Maybe 'Workload's Pods' is hard to understand. Does this mean "Pods assigned to the Workload" or "Pods in the Workload"?
Maybe it's worth expanding the terms a bit here.

Reviewer: Pods in the workload. Maybe we can add a link to Workload.

Author: Done

allows you to associate them with cluster nodes through labels and taints.

**Note**: If your the resources in your cluster are homogeneous, you can use 
an [empty ResourceFlavor](#empty-resourceflavor) instead of adding labels to custom ResourceFlavors.

A sample ResourceFlavor looks like the following:

```yaml
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: spot
nodeSelector:
  instance-type: spot
taints:
- effect: NoSchedule
  key: spot
  value: "true"
```

You can use the `.metadata.name` field to reference a ResourceFlavor from a
[ClusterQueue](/docs/concepts/cluster_queue.md) in the `.spec.resources[*].flavors[*].name` field.

## ResourceFlavor labels

**Requires Kubernetes 1.23 or newer**

To associate a ResourceFlavor with a subset of nodes of your cluster, you can
configure the `.nodeSelector` field with matching node labels that uniquely identify
the nodes. If you are using [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
(or equivalent controllers), make sure that the controller is configured to add those labels when
adding new nodes.

> To guarantee that the workload Pods run on the nodes associated to the flavor

Reviewer 3: Maybe we can uppercase workload to `Workload` across the document to mark it a special Object, just like ResourceFlavor and ClusterQueue.

To associate a ResourceFlavor with a subset of nodes of your cluster, you can
configure the `.nodeSelector` field with matching node labels that uniquely identify
the nodes. If you are using [cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
(or equivalent controllers), make sure that the controller is configured to add those labels when
adding new nodes.

To guarantee that the workload Pods run on the nodes associated to the flavor
that Kueue selected, Kueue performs the following
steps:

1. When admitting a workload, Kueue evaluates the
   [`.nodeSelector`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector)
   and [`.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity)
   fields in the PodSpecs of your [Workload](workload.md) against the
   ResourceFlavor labels.

> 

Reviewer: same paragraph should be good, so remove this empty line.

   `ResourceFlavors` that don't match the node affinity of the workload
   cannot be assigned to a workload's podSet.

2. Once the workload is admitted, Kueue adds the ResourceFlavor labels to the
  `.nodeSelector` of the underlying workload Pod templates. This occurs if the workload

>    didn't specify the `ResourceFlavor` labels already.

Reviewer: as part of its nodeSelector.

   For example, for a [batch/v1.Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/),
   Kueue adds the labels to the `.spec.template.spec.nodeSelector` field. This
   guarantees that the workload's Pods can only be scheduled on the nodes
   targeted by the flavor that Kueue assigned to the workload.

## ResourceFlavor taints

To restrict the usage of a ResourceFlavor, you can configure the `.taints` field

> with taints. These taints should typically match the taints of the Nodes associated with ResourceFlavor.

Reviewer: the* ResourceFlavor
